modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/RwTxFinishedHandler.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.tx.impl;
+
+import org.apache.ignite.internal.catalog.CatalogService;
+import org.apache.ignite.internal.hlc.HybridClock;
+import org.apache.ignite.internal.tx.TxStateMeta;
+import org.apache.ignite.internal.tx.message.RwTransactionsFinishedRequest;
+import org.apache.ignite.internal.tx.message.TxMessageGroup;
+import org.apache.ignite.internal.tx.message.TxMessagesFactory;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.network.NetworkMessageHandler;
+import org.jetbrains.annotations.Nullable;
+
+/** Auxiliary {@link RwTransactionsFinishedRequest} handler to encapsulate the logic associated with this handling. */
+class RwTxFinishedHandler {
```

Comment:
Naming is confusing. Same as for `RwTransactionsFinishedRequest` TxFinish request has another meaning in tx flow. I also doubt that given requests should be in tx module because it's not a part of tx protocol, should we move it to index module or something?

Reply:
Yes, now this is used by indexes, but for example, we can later easily adapt the logic for other algorithms, for example, for collecting statistics about transactions.
We could somehow move this into the index module, but it is important for us to ensure atomicity and consistency when processing messages. I don’t know how convenient and beautiful it is to provide this during the creation of a transaction and it looks cumbersome in reality.